File: apps/android/construction-android/app/src/main/java/com/constructionpro/app/features/projects/data/ProjectsRepository.kt
package com.constructionpro.app.features.projects.data

import com.apollographql.apollo3.ApolloClient
import com.apollographql.apollo3.api.Optional
import com.apollographql.apollo3.cache.normalized.FetchPolicy
import com.apollographql.apollo3.cache.normalized.fetchPolicy
import com.constructionpro.app.data.local.ProjectDao
import com.constructionpro.app.data.local.ProjectEntity
import com.constructionpro.app.graphql.GetProjectQuery
import com.constructionpro.app.graphql.GetProjectsQuery
import com.constructionpro.app.graphql.type.ProjectStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import java.time.LocalDate
import javax.inject.Inject
import javax.inject.Singleton

sealed class ProjectsResult<out T> {
    data class Success<T>(val data: T) : ProjectsResult<T>()
    data class Error(val message: String, val isOffline: Boolean = false) : ProjectsResult<Nothing>()
    data object Loading : ProjectsResult<Nothing>()
}

data class ProjectSummary(
    val id: String,
    val name: String,
    val description: String?,
    val status: String,
    val startDate: LocalDate?,
    val endDate: LocalDate?,
    val address: ProjectAddress,
    val dailyLogCount: Int,
    val documentCount: Int,
    val openIncidentCount: Int
)

data class ProjectDetail(
    val id: String,
    val name: String,
    val description: String?,
    val status: String,
    val startDate: LocalDate?,
    val endDate: LocalDate?,
    val address: ProjectAddress,
    val client: ClientInfo?,
    val team: List<TeamMember>,
    val dailyLogCount: Int,
    val documentCount: Int,
    val openIncidentCount: Int
)

data class ProjectAddress(
    val street: String?,
    val city: String?,
    val state: String?,
    val zipCode: String?,
    val country: String?,
    val latitude: Double?,
    val longitude: Double?,
    val formatted: String
)

data class ClientInfo(
    val id: String,
    val companyName: String,
    val contactName: String?,
    val email: String?,
    val phone: String?
)

data class TeamMember(
    val id: String,
    val name: String,
    val email: String?,
    val role: String,
    val avatarUrl: String?
)

@Singleton
class ProjectsRepository @Inject constructor(
    private val apolloClient: ApolloClient,
    private val projectDao: ProjectDao
) {

    fun getProjects(
        status: String? = null,
        search: String? = null,
        forceRefresh: Boolean = false
    ): Flow<ProjectsResult<List<ProjectSummary>>> = flow {
        emit(ProjectsResult.Loading)

        try {
            // First emit cached data
            val cached = if (search != null) {
                projectDao.search("%$search%")
            } else {
                projectDao.getAll()
            }

            if (cached.isNotEmpty()) {
                emit(ProjectsResult.Success(cached.map { it.toSummary() }))
            }

            // Then fetch from network
            val fetchPolicy = if (forceRefresh) FetchPolicy.NetworkOnly else FetchPolicy.CacheFirst

            val response = apolloClient.query(
                GetProjectsQuery(
                    status = Optional.presentIfNotNull(status?.let { ProjectStatus.safeValueOf(it) }),
                    search = Optional.presentIfNotNull(search),
                    page = Optional.present(1),
                    pageSize = Optional.present(50)
                )
            )
                .fetchPolicy(fetchPolicy)
                .execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                emit(ProjectsResult.Error(errorMessage))
                return@flow
            }

            val projects = response.data?.projects?.edges?.map { edge ->
                val node = edge.node
                ProjectSummary(
                    id = node.id,
                    name = node.name,
                    description = node.description,
                    status = node.status.name,
                    startDate = node.startDate?.toString()?.let { LocalDate.parse(it) },
                    endDate = node.endDate?.toString()?.let { LocalDate.parse(it) },
                    address = ProjectAddress(
                        street = node.address.street,
                        city = node.address.city,
                        state = node.address.state,
                        zipCode = node.address.zipCode,
                        country = null,
                        latitude = node.address.latitude,
                        longitude = node.address.longitude,
                        formatted = node.address.formatted
                    ),
                    dailyLogCount = node.dailyLogCount,
                    documentCount = node.documentCount,
                    openIncidentCount = node.openIncidentCount
                )
            } ?: emptyList()

            // Cache the results
            val entities = projects.map { project ->
                ProjectEntity(
                    id = project.id,
                    name = project.name,
                    status = project.status,
                    address = project.address.formatted,
                    updatedAt = System.currentTimeMillis()
                )
            }
            projectDao.insertAll(entities)

            emit(ProjectsResult.Success(projects))

        } catch (e: Exception) {
            // Network error - return cached data with offline flag
            val cached = if (search != null) {
                projectDao.search("%$search%")
            } else {
                projectDao.getAll()
            }

            if (cached.isNotEmpty()) {
                emit(ProjectsResult.Success(cached.map { it.toSummary() }))
            } else {
                emit(ProjectsResult.Error(e.message ?: "Network error", isOffline = true))
            }
        }
    }.flowOn(Dispatchers.IO)

    suspend fun getProject(id: String): ProjectsResult<ProjectDetail> = withContext(Dispatchers.IO) {
        try {
            val response = apolloClient.query(GetProjectQuery(id))
                .fetchPolicy(FetchPolicy.CacheFirst)
                .execute()

            if (response.hasErrors()) {
                val errorMessage = response.errors?.firstOrNull()?.message ?: "Unknown error"
                return@withContext ProjectsResult.Error(errorMessage)
            }

            val project = response.data?.project
                ?: return@withContext ProjectsResult.Error("Project not found")

            ProjectsResult.Success(
                ProjectDetail(
                    id = project.id,
                    name = project.name,
                    description = project.description,
                    status = project.status.name,
                    startDate = project.startDate?.toString()?.let { LocalDate.parse(it) },
                    endDate = project.endDate?.toString()?.let { LocalDate.parse(it) },
                    address = ProjectAddress(
                        street = project.address.street,
                        city = project.address.city,
                        state = project.address.state,
                        zipCode = project.address.zipCode,
                        country = project.address.country,
                        latitude = project.address.latitude,
                        longitude = project.address.longitude,
                        formatted = project.address.formatted
                    ),
                    client = project.client?.let { client ->
                        ClientInfo(
                            id = client.id,
                            companyName = client.companyName,
                            contactName = client.contactName,
                            email = client.email,
                            phone = client.phone
                        )
                    },
                    team = project.team.map { member ->
                        TeamMember(
                            id = member.id,
                            name = member.name,
                            email = member.email,
                            role = member.role.name,
                            avatarUrl = member.avatarUrl
                        )
                    },
                    dailyLogCount = project.dailyLogCount,
                    documentCount = project.documentCount,
                    openIncidentCount = project.openIncidentCount
                )
            )
        } catch (e: Exception) {
            ProjectsResult.Error(e.message ?: "Network error", isOffline = true)
        }
    }

    private fun ProjectEntity.toSummary() = ProjectSummary(
        id = id,
        name = name,
        description = null,
        status = status ?: "ACTIVE",
        startDate = null,
        endDate = null,
        address = ProjectAddress(
            street = null,
            city = null,
            state = null,
            zipCode = null,
            country = null,
            latitude = null,
            longitude = null,
            formatted = address ?: ""
        ),
        dailyLogCount = 0,
        documentCount = 0,
        openIncidentCount = 0
    )
}

File: apps/ios/ConstructionManager/ConstructionManager/Features/Projects/Data/ProjectsRepository.swift
//
//  ProjectsRepository.swift
//  ConstructionManager
//
//  Repository for Projects feature with GraphQL and offline support
//

import Foundation
import Combine

// MARK: - Result Types

enum ProjectsResult<T> {
    case success(T)
    case error(String, isOffline: Bool)
    case loading
}

// MARK: - Models

struct ProjectSummary: Identifiable, Codable {
    let id: String
    let name: String
    let description: String?
    let status: String
    let startDate: Date?
    let endDate: Date?
    let address: ProjectAddress
    let dailyLogCount: Int
    let documentCount: Int
    let drawingCount: Int
    let crewCount: Int
    let openIncidentCount: Int
    let createdAt: Date
    let updatedAt: Date
}

struct ProjectDetail: Identifiable, Codable {
    let id: String
    let name: String
    let description: String?
    let status: String
    let startDate: Date?
    let endDate: Date?
    let address: ProjectAddress
    let client: ClientInfo?
    let team: [TeamMemberInfo]
    let dailyLogCount: Int
    let documentCount: Int
    let drawingCount: Int
    let crewCount: Int
    let openIncidentCount: Int
    let createdAt: Date
    let updatedAt: Date
}

struct ProjectAddress: Codable {
    let street: String?
    let city: String?
    let state: String?
    let zipCode: String?
    let country: String?
    let latitude: Double?
    let longitude: Double?
    let formatted: String
}

struct ClientInfo: Codable {
    let id: String
    let companyName: String
    let contactName: String?
    let email: String?
    let phone: String?
}

struct TeamMemberInfo: Identifiable, Codable {
    let id: String
    let name: String
    let email: String?
    let role: String
    let avatarUrl: String?
}

// MARK: - Repository

@MainActor
class ProjectsRepository: ObservableObject {
    static let shared = ProjectsRepository()

    @Published private(set) var projects: [ProjectSummary] = []
    @Published private(set) var isLoading = false
    @Published private(set) var isUsingCachedData = false
    @Published var error: String?

    private let graphQLClient = GraphQLClient.shared
    private let networkMonitor = NetworkMonitor.shared
    private let offlineDataStore = OfflineDataStore.shared

    private var fetchTask: Task<Void, Never>?
    private var lastFetchTime: Date?
    private let minFetchInterval: TimeInterval = 2.0

    private init() {}

    // MARK: - Fetch Projects

    func fetchProjects(
        status: String? = nil,
        search: String? = nil,
        forceRefresh: Bool = false
    ) async {
        // Skip if recently fetched
        if !forceRefresh, let lastFetch = lastFetchTime,
           Date().timeIntervalSince(lastFetch) < minFetchInterval {
            return
        }

        // Deduplicate requests
        if let existingTask = fetchTask {
            await existingTask.value
            return
        }

        fetchTask = Task {
            await performFetch(status: status, search: search, forceRefresh: forceRefresh)
        }

        await fetchTask?.value
        fetchTask = nil
    }

    private func performFetch(
        status: String?,
        search: String?,
        forceRefresh: Bool
    ) async {
        isLoading = true
        error = nil
        isUsingCachedData = false
        defer { isLoading = false }

        // Check network connectivity
        guard networkMonitor.isConnected else {
            loadFromOfflineCache()
            return
        }

        // TODO: Once Apollo iOS code generation is complete, use:
        // let query = GetProjectsQuery(
        //     status: status.map { GraphQLNullable.some(ProjectStatus(rawValue: $0) ?? .active) } ?? .null,
        //     search: search.map { GraphQLNullable.some($0) } ?? .null,
        //     page: .some(1),
        //     pageSize: .some(50)
        // )
        //
        // graphQLClient.apollo.fetch(query: query, cachePolicy: forceRefresh ? .fetchIgnoringCacheData : .returnCacheDataElseFetch) { [weak self] result in
        //     switch result {
        //     case .success(let graphQLResult):
        //         if let errors = graphQLResult.errors, !errors.isEmpty {
        //             self?.error = errors.first?.message
        //             self?.loadFromOfflineCache()
        //             return
        //         }
        //
        //         if let projects = graphQLResult.data?.projects.edges {
        //             self?.projects = projects.map { $0.node.toProjectSummary() }
        //             self?.lastFetchTime = Date()
        //             self?.offlineDataStore.saveProjects(self?.projects ?? [])
        //         }
        //
        //     case .failure(let error):
        //         self?.error = error.localizedDescription
        //         self?.loadFromOfflineCache()
        //     }
        // }

        // Fallback to REST while Apollo is being set up
        await fetchProjectsViaREST(status: status, search: search)
    }

    /// Fallback to REST API
    private func fetchProjectsViaREST(status: String?, search: String?) async {
        do {
            var queryItems: [URLQueryItem] = []
            if let status = status {
                queryItems.append(URLQueryItem(name: "status", value: status))
            }
            if let search = search {
                queryItems.append(URLQueryItem(name: "search", value: search))
            }

            let response: ProjectsResponse = try await APIClient.shared.get(
                "/projects",
                queryItems: queryItems.isEmpty ? nil : queryItems
            )

            self.projects = response.projects.map { apiProject in
                ProjectSummary(
                    id: apiProject.id,
                    name: apiProject.name,
                    description: apiProject.description,
                    status: apiProject.status,
                    startDate: apiProject.startDate,
                    endDate: apiProject.endDate,
                    address: ProjectAddress(
                        street: nil,
                        city: nil,
                        state: nil,
                        zipCode: nil,
                        country: nil,
                        latitude: apiProject.gpsLatitude,
                        longitude: apiProject.gpsLongitude,
                        formatted: apiProject.address ?? ""
                    ),
                    dailyLogCount: apiProject.dailyLogCount ?? 0,
                    documentCount: apiProject.documentCount ?? 0,
                    drawingCount: apiProject.drawingCount ?? 0,
                    crewCount: apiProject.crewCount ?? 0,
                    openIncidentCount: 0,
                    createdAt: apiProject.createdAt ?? Date(),
                    updatedAt: apiProject.updatedAt ?? Date()
                )
            }
            self.lastFetchTime = Date()

            // Cache for offline use
            cacheProjects(projects)

        } catch {
            self.error = error.localizedDescription
            loadFromOfflineCache()
        }
    }

    // MARK: - Get Single Project

    func getProject(id: String) async -> ProjectsResult<ProjectDetail> {
        guard networkMonitor.isConnected else {
            // TODO: Load from cache
            return .error("No internet connection", isOffline: true)
        }

        // TODO: Use GraphQL once code generation is complete
        // let query = GetProjectQuery(id: id)
        // graphQLClient.apollo.fetch(query: query) { result in ... }

        do {
            let response: ProjectResponse = try await APIClient.shared.get("/projects/\(id)")
            let apiProject = response.project

            let detail = ProjectDetail(
                id: apiProject.id,
                name: apiProject.name,
                description: apiProject.description,
                status: apiProject.status,
                startDate: apiProject.startDate,
                endDate: apiProject.endDate,
                address: ProjectAddress(
                    street: nil,
                    city: nil,
                    state: nil,
                    zipCode: nil,
                    country: nil,
                    latitude: apiProject.gpsLatitude,
                    longitude: apiProject.gpsLongitude,
                    formatted: apiProject.address ?? ""
                ),
                client: apiProject.client.map {
                    ClientInfo(
                        id: $0.id,
                        companyName: $0.companyName,
                        contactName: $0.contactName,
                        email: nil,
                        phone: nil
                    )
                },
                team: [],
                dailyLogCount: apiProject.dailyLogCount ?? 0,
                documentCount: apiProject.documentCount ?? 0,
                drawingCount: apiProject.drawingCount ?? 0,
                crewCount: apiProject.crewCount ?? 0,
                openIncidentCount: 0,
                createdAt: apiProject.createdAt ?? Date(),
                updatedAt: apiProject.updatedAt ?? Date()
            )

            return .success(detail)

        } catch {
            return .error(error.localizedDescription, isOffline: false)
        }
    }

    // MARK: - Offline Support

    private func loadFromOfflineCache() {
        let cached = loadCachedProjects()
        if !cached.isEmpty {
            self.projects = cached
            self.isUsingCachedData = true
        } else {
            self.error = "No internet connection and no cached data available"
        }
    }

    private func cacheProjects(_ projects: [ProjectSummary]) {
        guard let data = try? JSONEncoder().encode(projects) else { return }

        let fileManager = FileManager.default
        guard let cachesDir = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else { return }

        let cacheFile = cachesDir.appendingPathComponent("cached_projects.json")
        try? data.write(to: cacheFile)
    }

    private func loadCachedProjects() -> [ProjectSummary] {
        let fileManager = FileManager.default
        guard let cachesDir = fileManager.urls(for: .cachesDirectory, in: .userDomainMask).first else { return [] }

        let cacheFile = cachesDir.appendingPathComponent("cached_projects.json")
        guard let data = try? Data(contentsOf: cacheFile),
              let projects = try? JSONDecoder().decode([ProjectSummary].self, from: data) else {
            return []
        }

        return projects
    }

    // MARK: - Search

    func searchProjects(query: String) -> [ProjectSummary] {
        guard !query.isEmpty else { return projects }
        return projects.filter {
            $0.name.localizedCaseInsensitiveContains(query) ||
            $0.address.formatted.localizedCaseInsensitiveContains(query)
        }
    }
}
